compiler/rustc_middle/src/mir/mono.rs

```diff
@@ -54,6 +54,33 @@ pub enum MonoItem<'tcx> {
     GlobalAsm(ItemId),
 }
 
+fn opt_incr_drop_glue_mode<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> InstantiationMode {
+    // Non-ADTs can't have a Drop impl. This case is mostly hit by closures whose captures require
+    // dropping.
+    let Some(adt_def) = ty.ty_adt_def() else {
+        return InstantiationMode::LocalCopy;
+    };
+
+    // Types that don't have a direct Drop impl, but have fields that require dropping.
+    let Some(dtor) = adt_def.destructor(tcx) else {
+        if adt_def.is_enum() {
```

Comment:
Do you know what motivates this choice of making enums act differently than structs? I mean, both of them have fields after all.

Reply:
I know something. I've added a comment that explains what I know and references the PR that landed the logic that I'm rewriting.